### PR TITLE
Adaptor: Add MySQL

### PR DESCRIPTION
**Pull request by atomicules**
I think this is at the point now where it's ready for review even if we don't merge yet.

There is obviously a LOT to review so will try to go over the main things worth looking at and considering:

- This MySQL adaptor was built using the Postgresql one as a basis
- You can follow the commit history to see how it was developed. But basically I copied the basic files over (`client.go`, `mysql.go`, `reader.go`, `session.go`, `writer.go`) from Postgresql (obviously renamed to `mysql.go`) and then added a test file at a time and used the tests to help me develop each file. Started with `client` first as that was easiest
- ~Does it make sense that we have all this [special handling in `reader_test.go`](https://github.com/compose/transporter/blob/9d6d829ed2c2c71688c4d7e427147d83c228ac1b/adaptor/mysql/reader_test.go#L134-L167) whereas Postgresql doesn't? I'm undecided. On the one hand I think this shouldn't be required as `reader.go` should do whatever is necessary, but on the other hand we don't necessarily specify how we are storing things at the Transporter level (i.e. compatibility layer between all adaptors, Golang native types). Maybe that should be in [`casifyValue` instead](https://github.com/compose/transporter/blob/9d6d829ed2c2c71688c4d7e427147d83c228ac1b/adaptor/mysql/reader.go#L201-L231)? Could do with another pair of eyes, etc on this~. In writing that comment I talked myself into changing things. See: https://github.com/compose/transporter/pull/515/commits/c3fabd279ce7ed1412a0d9d24332cf743fa8b272 but I still welcome someone else's opinion on this.
- It's worth looking through all the comments to see if that gives further things to think about. There are a lot of comments in `tailer.go`
- Is any of the resumemap stuff relevant? To Postgresql either?
- Uses [`go-mysql-org/go-mysql`](https://github.com/go-mysql-org/go-mysql/) as a driver over [`go-sql-driver/mysql`](https://github.com/go-sql-driver/mysql) because the former supports replication. Currently references a specific commit rather than a release tag as TLS support for the driver has only just been merged in.
- There are some `TODO`s left in the code, but I don't think these are things that need fixing *right now*. They are things that could be followed up with later. But you might disagree. Most are in `tailer.go`.
- Performance is probably similar to Postgresql. I.e. we aren't doing any bulk inserts yet.
- It has been tested with this [test dataset](https://www.mysqltutorial.org/mysql-sample-database.aspx) copying between two DBaaS providers MySQLs using TLS connections and with tailing.
- It runs `SET FOREIGN_KEY_CHECKS=0;` at the start of a session on the sink
- It requires the connecting user to be able to read the binlog for tailing. Some DBaaS providers might not provide this *right now*
- It's only been tested with a mysql sink and source. Unsure how it functions between other DB types. I don't think that should stop merging, but I could perhaps add a caveat to the mysql README about that.
- Regarding handling binlog rotation, testing has shown that the `replication/binlogsyncer` handles this
- Per Postgresql (and MongoDB?) need to create the destination database and table structure first. We _could_ handle this in Transporter, but per the other adaptors probably makes sense to handle that in some other place, e.g. Code that references Transporter.

Fixes: https://github.com/compose/transporter/issues/76

### Required for all PRs:

- [x] CHANGELOG.md updated (feel free to wait until changes have been reviewed by a maintainer)
- [x] README.md updated (if needed)

### CODE REVIEW

**Comment by atomicules** [comment]
Bah. `gofmt`ed everything, but still need to make some linting changes it seems.

**Comment by atomicules** [comment]
🎉 Checks passing.

**Comment by atomicules** [comment]
Checks are now failing. Github actions have switched from focal to jammy and weirdly that caused a failure. Fix here: https://github.com/compose/transporter/pull/516

File: adaptor/mysql/reader.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+							arrayRegexp := regexp.MustCompile("[[]]$")
+							if arrayRegexp.MatchString(columns[i][1]) {
+							} else {
+								docMap[columns[i][0]] = value
+							}
+						}
+					}
+					out <- doc{table: c, data: docMap}
+				}
+				log.With("db", db).With("table", c).Infoln("iterating complete")
```
Comment (atomicules) [inline adaptor/mysql/reader.go:193]: This log line seems to be causing an issue for Compose's internal use of Transporter. We might need to re-word it (even though it's the same as Postgresql and MongoDB).

Noting this down so it gets picked up in review.

Comment (atomicules) [reply adaptor/mysql/reader.go:194]: This has been resolved outside of this PR. I.e: issue was not with Transporter.

**Review by pete-may — APPROVED** [review]
LGTM!!!

- Tested `./transporter init mysql mysql` works
- Tested moving a couple rows between two Compose mysqls works
- Tested tailing between two mysqls works

Code looks really clean and you seem to have covered more things than I would ever have even imagined as a first pass at getting this working (handling spatial objects). So I say let's merge this monster. Can always fix bugs as they come up.

One suggestion would be to move your development notes from the mysql readme into a DEVELOPMENT_NOTES.md file

**Comment by atomicules** [comment]
Since we use squashed commits I'm keeping the branch, but will protect it, as otherwise all the useful commit history will be lost (i.e. if anyone else ever wants to use it as a basis for adding another adaptor).

**Pull request merged**